sw/device/tests/dif/dif_uart_test_mocked.cc

```diff
@@ -0,0 +1,734 @@
+   * other initialisation/configuration register access.
+   */
+  void DeviceResetExpectations(void) {
+    EXPECT_WRITE32(UART_CTRL_REG_OFFSET, 0);
+    EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0b00000011);
+    EXPECT_WRITE32(UART_OVRD_REG_OFFSET, 0);
+    EXPECT_WRITE32(UART_TIMEOUT_CTRL_REG_OFFSET, 0);
+    EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, 0);
+    EXPECT_WRITE32(UART_INTR_STATE_REG_OFFSET, kUint32Max);
+  }
+};
+
+class DifUartSendReceiveTest : public DifUartTest {
+ protected:
+  static const uint8_t kUartFifoSize_ = 32;
+  const std::array<uint8_t, kUartFifoSize_> kBytesSendDataArray_ = {
+      'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b',
+      'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'c', 'c',
+      'c', 'c', 'd', 'd', 'd', 'd', 'd', 'd', 'd', 'd',
+  };
+
+  std::array<uint8_t, kUartFifoSize_> kBytesReceiveDataArray_ = {};
+
+  /**
+   * Get the underlying uint8_t pointer to data of std::array "send".
+   */
+  const uint8_t *GetSendBytes(void) { return kBytesSendDataArray_.data(); }
+
+  /**
+   * Get the underlying uint8_t pointer to data of std::array "receive".
+   */
+  uint8_t *GetReceiveBytes(void) { return kBytesReceiveDataArray_.data(); }
+
+  /**
+   * Get UART FIFO size.
+   */
+  uint8_t GetFifoSize(void) { return kUartFifoSize_; }
+
+  /**
+   * Sets TX bytes expectations
+   *
+   * Every sent byte by the "send bytes" routine is expected to result in the
+   * STATUS read of 0 (FIFO not full), and write to WDATA.
+   */
+  void SetTxFifoExpectations(uint8_t num_elements) {
+    for (uint8_t i = 0; i < num_elements; ++i) {
+      uint32_t value = static_cast<uint32_t>(kBytesSendDataArray_[i]);
+      EXPECT_READ32(UART_STATUS_REG_OFFSET, 0);
+      EXPECT_WRITE32(UART_WDATA_REG_OFFSET, value);
+    }
+  }
+
+  /**
+   * Sets RX bytes expectations
+   *
+   * Every read byte by the "receive bytes" routine is expected to result in the
+   * STATUS read of 0 (FIFO not empty), and read of RDATA.
+   */
+  void SetRxFifoExpectations(uint8_t num_elements) {
+    for (uint8_t i = 0; i < num_elements; ++i) {
+      uint32_t value = static_cast<uint32_t>(kBytesSendDataArray_[i]);
+      EXPECT_READ32(UART_STATUS_REG_OFFSET, 0);
+      EXPECT_READ32(UART_RDATA_REG_OFFSET, value);
+    }
+  }
+
+  /**
+   * Compares reference array, which is used to set read expectations, with the
+   * array of read bytes. When less than reference array size elements have been
+   * requested, check that remaining bytes of the received array remain unset.
+   */
+  void CompareSendReceiveExpectations(uint8_t num_elements) {
+    assert(num_elements <= kBytesSendDataArray_.size());
+
+    for (uint8_t i = 0; i < num_elements; ++i) {
+      EXPECT_EQ(kBytesSendDataArray_[i], kBytesReceiveDataArray_[i]);
+    }
+
+    // Check that nothing that hasn't been requested, is written.
+    for (uint8_t i = num_elements; i < kBytesSendDataArray_.size(); ++i) {
+      EXPECT_EQ(kBytesReceiveDataArray_[i], 0);
+    }
```

Comment:
C++ implements equality for various containers. Consider using std::vector, which will allow you to compress these checks into a single EXPECT_EQ. I also suspect there is an EXPECT_THAT matcher that can help you.
https://github.com/google/googletest/blob/master/googlemock/docs/cheat_sheet.md

Reply:
Done.